gcovr/workers.py

```diff
@@ -0,0 +1,144 @@
+# -*- coding:utf-8 -*-
+
+# This file is part of gcovr <http://gcovr.com/>.
+#
+# Copyright 2013-2018 the gcovr authors
+# Copyright 2013 Sandia Corporation
+# This software is distributed under the BSD license.
+
+
+from threading import Thread, Condition
+from contextlib import contextmanager
+
+import sys
+if sys.version_info[0] >= 3:
+    from queue import Queue, Empty
+else:
+    from Queue import Queue, Empty
+
+
+class LockedDirectories(object):
+    """
+    Class that keeps a list of locked directories
+    """
+    def __init__(self):
+        self.dirs = set()
+        self.cv = Condition()
+
+    def run_in(self, dir_):
+        """
+        Start running in the directory and lock it
+        """
+        self.cv.acquire()
+        while dir_ in self.dirs:
+            self.cv.wait()
+        self.dirs.add(dir_)
+        self.cv.release()
+
+    def done(self, dir_):
+        """
+        Finished with the directory, unlock it
+        """
+        self.cv.acquire()
+        self.dirs.remove(dir_)
+        self.cv.notify_all()
+        self.cv.release()
+
+
+@contextmanager
+def locked_directory(dir_):
+    """
+    Context for doing something in a locked directory
+    """
+    locked_directory.global_object.run_in(dir_)
+    yield
+    locked_directory.global_object.done(dir_)
+
+
+locked_directory.global_object = LockedDirectories()
+
+
+class WorkThread(Thread):
+    """
+    The work thread class continuously gets work and
+    completes it
+    """
+    def __init__(self, pool):
+        """
+        Initialise with a reference to the pool object
+        which houses the queue
+        """
+        super(WorkThread, self).__init__()
+        import tempfile
+        self.pool = pool
+        self.workdir = tempfile.mkdtemp()
+
+    def run(self):
+        """
+        Run until the queue is empty
+        """
+        while True:
+            try:
+                work, args, kwargs = self.pool.get()
+            except Empty:
+                break
+            kwargs['workdir'] = self.workdir
+            work(*args, **kwargs)
+
+    def close(self):
+        """
+        Empty the working directory
+        """
+        import shutil
+
+        # On Windows the files may still be in use. This
+        # is unlikely, the files are small, and are in a
+        # temporary directory so we can skip this.
+        shutil.rmtree(self.workdir, ignore_errors=True)
```

Comment:
Why would the files still be in use on Windows ? This doesn't seem right. Any pointers ?

Reply:
Indeed it does not seem right - as these are just parsed by python and not touched by gcov. I think this may be able to go once I've cured the (potential) issue with deletion of the *.gcov files. I've set up a local Windows test machine now so I should get this done today.